src/powershell/Program.cs

```diff
@@ -19,7 +63,458 @@ public sealed class ManagedPSEntry
+        /// Gets what the length of the given string will be if it's
+        /// quote escaped for /bin/sh.
+        /// </summary>
+        /// <param name="str">The string to quote escape.</param>
+        /// <returns>The length of the string when it's quote escaped.</returns>
+        private static int GetQuotedPathLength(string str)
+        {
+            int length = 2;
+            foreach (char c in str)
+            {
+                length++;
+                if (c == '\'') { length++; }
+            }
+
+            return length;
+        }
+
+        /// <summary>
+        /// Implements a SpanAction&lt;T&gt; for string.Create()
+        /// that builds the shell invocation for the login pwsh session.
+        /// </summary>
+        /// <param name="strBuf">The buffer of the string to be created.</param>
+        /// <param name="invocationInfo">Information used to build the required string.</param>
+        private static void CreatePwshInvocation(
+            Span<char> strBuf,
+            (string path, int quotedLength) invocationInfo)
+        {
+            // "exec "
+            string prefix = "exec ";
+            prefix.AsSpan().CopyTo(strBuf);
+
+            // The quoted path to pwsh, like "'/opt/microsoft/powershell/7/pwsh'"
+            int i = prefix.Length;
+            Span<char> pathSpan = strBuf.Slice(i, invocationInfo.quotedLength);
+            QuoteAndWriteToSpan(invocationInfo.path, pathSpan);
+            i += invocationInfo.quotedLength;
+
+            // ' "$@"' the argument vector splat to pass pwsh arguments through
+            string suffix = " \"$@\"";
+            Span<char> bufSuffix = strBuf.Slice(i);
+            suffix.AsSpan().CopyTo(bufSuffix);
+        }
+
+        /// <summary>
+        /// Quotes (and sh quote escapes) a string and writes it to the given span.
+        /// </summary>
+        /// <param name="arg">The string to quote.</param>
+        /// <param name="span">The span to write to.</param>
+        private static void QuoteAndWriteToSpan(string arg, Span<char> span)
+        {
+            span[0] = '\'';
+
+            int i = 0;
+            int j = 1;
+            for (; i < arg.Length; i++, j++)
+            {
+                char c = arg[i];
+
+                if (c == '\'')
+                {
+                    // /bin/sh quote escaping uses backslashes
+                    span[j] = '\\';
+                    j++;
+                }
+
+                span[j] = c;
+            }
+
+            span[j] = '\'';
+        }
+
+        /// <summary>
+        /// If the given exit code is negative, throws a StartupException.
+        /// </summary>
+        /// <param name="call">The native call that was attempted.</param>
+        /// <param name="code">The exit code it returned.</param>
+        private static void ThrowOnFailure(string call, int code)
+        {
+            if (code < 0)
+            {
+                code = Marshal.GetLastWin32Error();
```

Comment:
I doubt this will have any effect. According to the [doc](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getlastwin32error?view=netcore-3.0), this works only if the unmanaged function can call the [`SetLastError`](https://docs.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-setlasterror) Windows API.

Reply:
It looks like I need to indicate `SetLastError` in the `DllImport` but otherwise, [I think this should work](https://developers.redhat.com/blog/2019/03/25/using-net-pinvoke-for-linux-system-functions/)

Reply:
The MSDN document about `GetLastWin32Error()` needs to be updated